explorer/interpreter/type_checker.cpp

```diff
@@ -662,6 +663,21 @@ auto TypeChecker::ArgumentDeduction(
       if (std::find(bindings_to_deduce.begin(), bindings_to_deduce.end(),
                     &var_type.binding()) != bindings_to_deduce.end()) {
         auto [it, success] = deduced.insert({&var_type.binding(), arg});
+
+        const Value* substituted_param_type = Substitute(deduced, param);
+        const auto& binding = cast<VariableType>(*param).binding();
+        if (binding.has_static_type()) {
+          if (!IsTypeOfType(&binding.static_type())) {
+            if (!IsImplicitlyConvertible(substituted_param_type,
+                                         arg, impl_scope, false)) {
+              return CompilationError(source_loc)
+                     << "cannot convert deduced value " << *arg
+                     << " for " << binding.name()
+                     << " to parameter type " << *substituted_param_type;
+            }
+          }
+        }
+
```

Comment:
```suggestion
      const auto& binding = var_type.binding();
      if (std::find(bindings_to_deduce.begin(), bindings_to_deduce.end(),
                    &binding) != bindings_to_deduce.end()) {
        if (binding.has_static_type()) {
          const auto* binding_type = Substitute(deduced, &binding.static_type());
          if (!IsTypeOfType(binding_type) &&
              !IsImplicitlyConvertible(
                  arg, binding_type, impl_scope,
                  /*allow_user_defined_conversions*/ false)) {
            return CompilationError(source_loc)
                   << "cannot convert deduced value " << *arg << " for parameter "
                   << binding.name() << " to parameter type "
                   << *binding_type;
          }
        }

        auto [it, success] = deduced.insert({&binding, arg});
```

Main change here is substituting into the type of the binding rather than into the `VariableType` value. I also reversed the arguments to `IsImplicitlyConvertible`, which I got wrong in a previous suggested edit.